File: web/src/ivfpq/pq.ts
/**
 * Product Quantization (PQ) implementation for browser-side inference using ONNX models.
 *
 * This module provides TypeScript bindings for Product Quantization ONNX models
 * exported from the Python training pipeline. It focuses on using the trained
 * ONNX models for encoding embeddings and computing asymmetric distances.
 *
 * Based on the paper "Product Quantization for Nearest Neighbor Search" by Jégou et al.
 */

import * as ort from 'onnxruntime-web'

/**
 * Product Quantization metadata structure.
 */
export interface PQMetadata {
  d: number // Input vector dimension
  m: number // Number of subquantizers
  k: number // Number of centroids per subquantizer
  d_sub: number // Dimension of each subspace (d/m)
  compression_ratio: number
  codebooks_shape: number[]
  codebooks_size: number
  training_samples: number
  max_iterations: number
  version: string
}

/**
 * Product Quantization system using ONNX models.
 */
export class ProductQuantizer {
  public readonly metadata: PQMetadata
  
  private codebooks: Float32Array | null = null
  private encodeSession: ort.InferenceSession | null = null
  private decodeSession: ort.InferenceSession | null = null
  private distanceSession: ort.InferenceSession | null = null
  private distanceBaseSession: ort.InferenceSession | null = null

  constructor(metadata: PQMetadata) {
    this.metadata = metadata
    
    if (metadata.d % metadata.m !== 0) {
      throw new Error(`Input dimension ${metadata.d} must be divisible by number of subquantizers ${metadata.m}`)
    }

    console.log(
      `Initialized ProductQuantizer: d=${metadata.d}, m=${metadata.m}, k=${metadata.k}, d_sub=${metadata.d_sub}`
    )
    console.log(`Compression ratio: ${metadata.compression_ratio}x`)
  }

  /**
   * Load codebooks from a Float32Array.
   * Expected format: Float32Array with shape [m, k, d_sub] flattened.
   */
  loadCodebooks(codebooks: Float32Array): void {
    const expectedSize = this.metadata.codebooks_size
    if (codebooks.length !== expectedSize) {
      throw new Error(`Codebook size mismatch: expected ${expectedSize}, got ${codebooks.length}`)
    }
    this.codebooks = codebooks
    console.log(`Loaded codebooks with shape [${this.metadata.m}, ${this.metadata.k}, ${this.metadata.d_sub}]`)
  }

  /**
   * Load ONNX model for encoding vectors to PQ codes.
   */
  async loadEncoder(modelPath: string): Promise<void> {
    try {
      this.encodeSession = await ort.InferenceSession.create(modelPath, {
        executionProviders: ['wasm'],
        logSeverityLevel: 3, // Warning level
      })
      console.log(`Loaded PQ encoder from ${modelPath}`)
    } catch (error) {
      console.error('Failed to load PQ encoder:', error)
      throw error
    }
  }

  /**
   * Load ONNX model for decoding PQ codes to vectors.
   */
  async loadDecoder(modelPath: string): Promise<void> {
    try {
      this.decodeSession = await ort.InferenceSession.create(modelPath, {
        executionProviders: ['wasm'],
        logSeverityLevel: 3,
      })
      console.log(`Loaded PQ decoder from ${modelPath}`)
    } catch (error) {
      console.error('Failed to load PQ decoder:', error)
      throw error
    }
  }

  /**
   * Load ONNX model for distance computation with top-k.
   */
  async loadDistanceModel(modelPath: string): Promise<void> {
    try {
      this.distanceSession = await ort.InferenceSession.create(modelPath, {
        executionProviders: ['wasm'],
        logSeverityLevel: 3,
      })
      console.log(`Loaded PQ distance model from ${modelPath}`)
    } catch (error) {
      console.error('Failed to load PQ distance model:', error)
      throw error
    }
  }

  /**
   * Load ONNX model for distance computation without top-k.
   */
  async loadDistanceBaseModel(modelPath: string): Promise<void> {
    try {
      this.distanceBaseSession = await ort.InferenceSession.create(modelPath, {
        executionProviders: ['wasm'],
        logSeverityLevel: 3,
      })
      console.log(`Loaded PQ distance base model from ${modelPath}`)
    } catch (error) {
      console.error('Failed to load PQ distance base model:', error)
      throw error
    }
  }

  /**
   * Encode vectors using the ONNX encoder model.
   *
   * @param vectors - Input vectors as Float32Array with shape [N, d]
   * @returns PQ codes as Uint8Array with shape [N, m]
   */
  async encode(vectors: Float32Array): Promise<Uint8Array> {
    if (!this.encodeSession) {
      throw new Error('PQ encoder model not loaded. Call loadEncoder() first.')
    }
    if (!this.codebooks) {
      throw new Error('Codebooks not loaded. Call loadCodebooks() first.')
    }

    const batchSize = vectors.length / this.metadata.d

    // Create input tensors
    const embeddingsTensor = new ort.Tensor('float32', vectors, [batchSize, this.metadata.d])
    const codebooksTensor = new ort.Tensor('float32', this.codebooks, [
      this.metadata.m, 
      this.metadata.k, 
      this.metadata.d_sub
    ])

    // Run inference
    const outputs = await this.encodeSession.run({
      embeddings: embeddingsTensor,
      codebooks: codebooksTensor
    })

    // Extract codes (should be int64)
    const codes = outputs.codes.data as BigInt64Array

    // Convert to Uint8Array (assuming codes are in valid range [0, k-1])
    const uint8Codes = new Uint8Array(codes.length)
    for (let i = 0; i < codes.length; i++) {
      uint8Codes[i] = Number(codes[i])
    }

    return uint8Codes
  }

  /**
   * Decode PQ codes using the ONNX decoder model.
   *
   * @param codes - PQ codes as Uint8Array with shape [N, m]
   * @returns Decoded vectors as Float32Array with shape [N, d]
   */
  async decode(codes: Uint8Array): Promise<Float32Array> {
    if (!this.decodeSession) {
      throw new Error('PQ decoder model not loaded. Call loadDecoder() first.')
    }
    if (!this.codebooks) {
      throw new Error('Codebooks not loaded. Call loadCodebooks() first.')
    }

    const batchSize = codes.length / this.metadata.m

    // Convert codes to int64 array for ONNX
    const int64Codes = new Array(codes.length)
    for (let i = 0; i < codes.length; i++) {
      int64Codes[i] = codes[i]
    }

    // Create input tensors
    const codesTensor = new ort.Tensor('int64', int64Codes, [batchSize, this.metadata.m])
    const codebooksTensor = new ort.Tensor('float32', this.codebooks, [
      this.metadata.m, 
      this.metadata.k, 
      this.metadata.d_sub
    ])

    // Run inference
    const outputs = await this.decodeSession.run({
      codes: codesTensor,
      codebooks: codebooksTensor
    })

    return outputs.embeddings.data as Float32Array
  }

  /**
   * Find k nearest neighbors using the ONNX distance model with top-k.
   *
   * @param queryVector - Query vector as Float32Array with shape [d]
   * @param referenceCodes - PQ codes for reference vectors as Uint8Array with shape [N, m]
   * @returns Indices of k nearest neighbors
   */
  async findNearestNeighbors(
    queryVector: Float32Array,
    referenceCodes: Uint8Array
  ): Promise<number[]> {
    if (!this.distanceSession) {
      throw new Error('PQ distance model not loaded. Call loadDistanceModel() first.')
    }
    if (!this.codebooks) {
      throw new Error('Codebooks not loaded. Call loadCodebooks() first.')
    }

    if (queryVector.length !== this.metadata.d) {
      throw new Error(
        `Query vector dimension ${queryVector.length} doesn't match expected ${this.metadata.d}`
      )
    }

    const numReferences = referenceCodes.length / this.metadata.m

    // Convert reference codes to int64 array for ONNX
    const int64RefCodes = new Array(referenceCodes.length)
    for (let i = 0; i < referenceCodes.length; i++) {
      int64RefCodes[i] = referenceCodes[i]
    }

    // Create input tensors
    const queryTensor = new ort.Tensor('float32', queryVector, [this.metadata.d])
    const refCodesTensor = new ort.Tensor('int64', int64RefCodes, [numReferences, this.metadata.m])
    const codebooksTensor = new ort.Tensor('float32', this.codebooks, [
      this.metadata.m, 
      this.metadata.k, 
      this.metadata.d_sub
    ])

    // Run inference
    const outputs = await this.distanceSession.run({
      query: queryTensor,
      reference_codes: refCodesTensor,
      codebooks: codebooksTensor
    })

    // Extract indices
    const indices = outputs.indices.data as BigInt64Array
    const result = new Array(indices.length)
    for (let i = 0; i < indices.length; i++) {
      result[i] = Number(indices[i])
    }

    return result
  }

  /**
   * Compute distances using the ONNX distance base model (no top-k).
   *
   * @param queryVector - Query vector as Float32Array with shape [d]
   * @param referenceCodes - PQ codes for reference vectors as Uint8Array with shape [N, m]
   * @returns Squared distances as Float32Array with shape [N]
   */
  async computeDistances(
    queryVector: Float32Array,
    referenceCodes: Uint8Array
  ): Promise<Float32Array> {
    if (!this.distanceBaseSession) {
      throw new Error('PQ distance base model not loaded. Call loadDistanceBaseModel() first.')
    }
    if (!this.codebooks) {
      throw new Error('Codebooks not loaded. Call loadCodebooks() first.')
    }

    if (queryVector.length !== this.metadata.d) {
      throw new Error(
        `Query vector dimension ${queryVector.length} doesn't match expected ${this.metadata.d}`
      )
    }

    const numReferences = referenceCodes.length / this.metadata.m

    // Convert reference codes to int64 array for ONNX
    const int64RefCodes = new Array(referenceCodes.length)
    for (let i = 0; i < referenceCodes.length; i++) {
      int64RefCodes[i] = referenceCodes[i]
    }

    // Create input tensors
    const queryTensor = new ort.Tensor('float32', queryVector, [this.metadata.d])
    const refCodesTensor = new ort.Tensor('int64', int64RefCodes, [numReferences, this.metadata.m])
    const codebooksTensor = new ort.Tensor('float32', this.codebooks, [
      this.metadata.m, 
      this.metadata.k, 
      this.metadata.d_sub
    ])

    // Run inference
    const outputs = await this.distanceBaseSession.run({
      query: queryTensor,
      reference_codes: refCodesTensor,
      codebooks: codebooksTensor
    })

    return outputs.distances.data as Float32Array
  }

  /**
   * Compute reconstruction error statistics.
   *
   * @param originalVectors - Original vectors as Float32Array with shape [N, d]
   * @param codes - PQ codes as Uint8Array with shape [N, m]
   * @returns Object with MSE and relative error statistics
   */
  async computeReconstructionError(
    originalVectors: Float32Array,
    codes: Uint8Array
  ): Promise<{
    mse: number
    relativeError: number
    compressionRatio: number
  }> {
    const decoded = await this.decode(codes)
    const batchSize = codes.length / this.metadata.m

    let totalSquaredError = 0
    let totalOriginalSquaredNorm = 0

    for (let n = 0; n < batchSize; n++) {
      for (let i = 0; i < this.metadata.d; i++) {
        const idx = n * this.metadata.d + i
        const original = originalVectors[idx]
        const reconstructed = decoded[idx]
        const error = original - reconstructed

        totalSquaredError += error * error
        totalOriginalSquaredNorm += original * original
      }
    }

    const mse = totalSquaredError / (batchSize * this.metadata.d)
    const relativeError = Math.sqrt(totalSquaredError / totalOriginalSquaredNorm)

    return {
      mse,
      relativeError,
      compressionRatio: this.metadata.compression_ratio,
    }
  }
}

/**
 * Load codebooks from a binary file.
 * This assumes the file contains a flattened Float32Array with the codebooks.
 */
export async function loadCodebooksFromFile(path: string): Promise<Float32Array> {
  const response = await fetch(path)
  if (!response.ok) {
    throw new Error(`Failed to load codebooks from ${path}: ${response.statusText}`)
  }

  const buffer = await response.arrayBuffer()
  return new Float32Array(buffer)
}

/**
 * Load PQ model metadata from JSON file.
 */
export async function loadPQMetadata(metadataPath: string): Promise<PQMetadata> {
  const response = await fetch(metadataPath)
  if (!response.ok) {
    throw new Error(`Failed to load metadata from ${metadataPath}: ${response.statusText}`)
  }

  return await response.json()
}

/**
 * Load complete PQ model (metadata + codebooks + ONNX models).
 */
export async function loadPQModel(
  basePath: string, 
  options: {
    loadEncoder?: boolean
    loadDecoder?: boolean
    loadDistance?: boolean
    loadDistanceBase?: boolean
  } = {}
): Promise<ProductQuantizer> {
  // Default to loading all models
  const {
    loadEncoder = true,
    loadDecoder = true,
    loadDistance = true,
    loadDistanceBase = true
  } = options

  // Load metadata
  const metadata = await loadPQMetadata(`${basePath}/pq_metadata.json`)

  // Create PQ instance
  const pq = new ProductQuantizer(metadata)

  // Load codebooks
  const codebooks = await loadCodebooksFromFile(`${basePath}/pq_codebooks.bin`)
  pq.loadCodebooks(codebooks)

  // Load ONNX models as requested
  if (loadEncoder) {
    await pq.loadEncoder(`${basePath}/pq_encode.onnx`)
  }

  if (loadDecoder) {
    await pq.loadDecoder(`${basePath}/pq_decode.onnx`)
  }

  if (loadDistance) {
    await pq.loadDistanceModel(`${basePath}/pq_distance.onnx`)
  }

  if (loadDistanceBase) {
    await pq.loadDistanceBaseModel(`${basePath}/pq_distance_base.onnx`)
  }

  console.log(`Loaded PQ model from ${basePath}:`)
  console.log(`  d=${metadata.d}, m=${metadata.m}, k=${metadata.k}, d_sub=${metadata.d_sub}`)

  return pq
}

/**
 * Utility function to validate vector dimensions.
 */
export function validateVectorDimensions(vectors: Float32Array, expectedDim: number): void {
  if (vectors.length % expectedDim !== 0) {
    throw new Error(`Array length ${vectors.length} is not divisible by dimension ${expectedDim}`)
  }
}

/**
 * Utility function to create test data for PQ models.
 */
export function generateTestVectors(numVectors: number, dimension: number, seed?: number): Float32Array {
  // Simple deterministic random number generator if seed is provided
  const rng = seed !== undefined ? (() => {
    let currentSeed = seed
    return () => {
      currentSeed = (currentSeed * 9301 + 49297) % 233280
      return currentSeed / 233280
    }
  })() : Math.random

  const vectors = new Float32Array(numVectors * dimension)
  for (let i = 0; i < vectors.length; i++) {
    vectors[i] = (rng() - 0.5) * 2 // Random values between -1 and 1
  }
  return vectors
}

/**
 * Utility function to create test PQ codes.
 */
export function generateTestCodes(numVectors: number, m: number, k: number, seed?: number): Uint8Array {
  const rng = seed !== undefined ? (() => {
    let currentSeed = seed
    return () => {
      currentSeed = (currentSeed * 9301 + 49297) % 233280
      return currentSeed / 233280
    }
  })() : Math.random

  const codes = new Uint8Array(numVectors * m)
  for (let i = 0; i < codes.length; i++) {
    codes[i] = Math.floor(rng() * k)
  }
  return codes
}

File: ivfpq/python/src/ivfpq/pq.py
"""
Product Quantization (PQ) implementation with optimized ONNX export.

This module provides PyTorch models for Product Quantization including:
- Encode: quantizes embeddings using learned codebooks
- Decode: reconstructs embeddings from PQ codes
- Distance: computes asymmetric distances for efficient kNN search

Based on "Product Quantization for Nearest Neighbor Search" by Jégou et al.
"""

import torch
import torch.nn as nn
import numpy as np
from typing import Tuple, Dict
from pathlib import Path
import json
import logging
from sklearn.cluster import KMeans
from tqdm import tqdm

logger = logging.getLogger(__name__)


class PQ(nn.Module):
    """
    Product Quantization encoder/decoder module.

    Product Quantization decomposes high-dimensional vectors into multiple lower-dimensional subvectors, each quantized independently using k-means.
    This allows for significant compression while maintaining reasonable
    approximation quality for search.

    Args:
        d: Input vector dimension
        m: Number of subspaces (subquantizers)
        k: Number of centroids per subspace (codebook size)

    The input dimension d must be divisible by m.
    Each subspace has dimension d/m and k centroids.
    """

    def __init__(self, d: int, m: int, k: int):
        super().__init__()

        if d % m != 0:
            raise ValueError(
                f"Input dimension {d} must be divisible by number of subquantizers {m}"
            )

        self.d = d  # Input dimension
        self.m = m  # Number of subquantizers
        self.k = k  # Number of centroids per subquantizer
        self.d_sub = d // m  # Dimension of each subspace

        # Codebooks: [m, k, d_sub] - m codebooks, each with k centroids of dimension d_sub
        self.codebooks = nn.Parameter(
            torch.randn(m, k, self.d_sub), requires_grad=False
        )

        logger.info(
            f"Initialized PQ: d={self.d}, m={self.m}, k={self.k}, d_sub={self.d_sub} for a compression ratio of {self.d * 32 / (self.m * 8)}x"
        )

        logger.info(
            f"Codebooks shape: {self.codebooks.shape} with dtype {self.codebooks.dtype} and storage size {self.codebooks.numel() * self.codebooks.element_size() / (1024 ** 2):.2f} MB"
        )

        # Whether the quantizer has been trained
        self.is_trained = False

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        """
        Encode vectors using product quantization.

        Args:
            x: Input vectors of shape [N, d]

        Returns:
            PQ codes of shape [N, m] with integer values in [0, k-1]
        """
        if not self.is_trained:
            raise RuntimeError("PQmust be trained before encoding")

        batch_size = x.shape[0]

        # Split input into subvectors: [N, m, d_sub]
        x_split = x.view(batch_size, self.m, self.d_sub)

        # Find closest centroids for each subvector
        codes = torch.zeros(batch_size, self.m, dtype=torch.long, device=x.device)

        for i in range(self.m):
            # Compute distances to all centroids in this subspace
            sub_vectors = x_split[:, i]  # [N, d_sub]
            centroids = self.codebooks[i]  # [k, d_sub]

            # Compute squared Euclidean distances using broadcasting (ONNX-friendly)
            sub_vectors_expanded = sub_vectors.unsqueeze(1)  # [N, 1, d_sub]
            centroids_expanded = centroids.unsqueeze(0)  # [1, k, d_sub]

            # Compute squared distances: [N, k]
            diff = sub_vectors_expanded - centroids_expanded  # [N, k, d_sub]
            distances = torch.sum(diff * diff, dim=2)  # [N, k]

            # Find nearest centroid indices
            codes[:, i] = torch.argmin(distances, dim=1)

        return codes

    def decode(self, codes: torch.Tensor) -> torch.Tensor:
        """
        Decode PQ codes back to approximate vectors.

        Args:
            codes: PQ codes of shape [N, m] with integer values in [0, k-1]

        Returns:
            Decoded vectors of shape [N, d]
        """
        if not self.is_trained:
            raise RuntimeError("PQmust be trained before decoding")

        batch_size = codes.shape[0]

        # Reconstruct vectors by looking up centroids
        decoded = torch.zeros(
            batch_size, self.d, device=codes.device, dtype=self.codebooks.dtype
        )

        for i in range(self.m):
            start_idx = i * self.d_sub
            end_idx = (i + 1) * self.d_sub

            # Look up centroids for this subspace (ensure codes are long type)
            decoded[:, start_idx:end_idx] = self.codebooks[i][codes[:, i].long()]

        return decoded

    def train_pq(
        self,
        vectors: torch.Tensor,
        n_iterations: int = 50,
    ) -> None:
        """
        Train the product quantizer using scikit-learn k-means clustering.

        Args:
            vectors: Training vectors of shape [N, d]
            n_iterations: Number of k-means iterations per subspace
        """
        if vectors.shape[1] != self.d:
            raise ValueError(
                f"Vector dimension {vectors.shape[1]} doesn't match expected {self.d}"
            )

        n_vectors = vectors.shape[0]

        # Split vectors into subvectors
        vectors_split = vectors.view(n_vectors, self.m, self.d_sub)

        with torch.no_grad():
            # Progress bar for subquantizers
            subquantizer_pbar = tqdm(range(self.m), desc="Training subquantizers")

            for i in subquantizer_pbar:
                sub_vectors = vectors_split[:, i]  # [N, d_sub]

                # Adjust n_clusters if we have fewer samples than clusters
                n_samples = sub_vectors.shape[0]
                n_clusters_actual = min(self.k, n_samples)

                # Use scikit-learn k-means
                kmeans = KMeans(
                    n_clusters=n_clusters_actual,
                    max_iter=n_iterations,
                    random_state=42 + i,
                    n_init=1,  # Only run once since we set the random state
                )

                # Convert to numpy for sklearn
                sub_vectors_np = sub_vectors.cpu().numpy()
                kmeans.fit(sub_vectors_np)
                centroids_actual = torch.from_numpy(kmeans.cluster_centers_).float()

                # Store learned centroids, padding with zeros if necessary
                if n_clusters_actual < self.k:
                    centroids = torch.zeros(self.k, self.d_sub)
                    centroids[:n_clusters_actual] = centroids_actual
                else:
                    centroids = centroids_actual

                self.codebooks.data[i] = centroids

                # Check cluster utilization
                assignments = kmeans.labels_
                unique_assignments = np.unique(assignments)
                utilization = len(unique_assignments)
                subquantizer_pbar.set_postfix(clusters_used=f"{utilization}/{self.k}")

            subquantizer_pbar.close()

        self.is_trained = True

    def save(self, path: Path) -> None:
        """Save the trained quantizer to disk in binary format."""
        if not self.is_trained:
            raise RuntimeError("Cannot save untrained quantizer")

        path = Path(path)
        path.mkdir(exist_ok=True, parents=True)

        # Save metadata as JSON
        metadata = {
            "d": self.d,
            "m": self.m,
            "k": self.k,
            "d_sub": self.d_sub,
            "compression_ratio": float((self.d * 32) / (self.m * 8)),
            "codebooks_shape": [self.m, self.k, self.d_sub],
            "codebooks_size": int(self.m * self.k * self.d_sub),
            "version": "1.0",
        }

        with open(path / "pq_metadata.json", "w") as f:
            json.dump(metadata, f, indent=2)

        # Save codebooks as binary file
        codebooks_np = self.codebooks.data.cpu().numpy().astype(np.float32)
        codebooks_np.tofile(path / "pq_codebooks.bin")

        logger.info(f"Saved binary codebooks to {path / 'pq_codebooks.bin'}")
        logger.info(f"Metadata saved to {path / 'pq_metadata.json'}")

    @classmethod
    def load(cls, path: Path) -> "PQ":
        """Load a trained quantizer from disk."""
        path = Path(path)

        # Load metadata
        with open(path / "pq_metadata.json", "r") as f:
            metadata = json.load(f)

        # Create quantizer
        quantizer = cls(metadata["d"], metadata["m"], metadata["k"])

        # Load codebooks from binary file
        codebooks_np = np.fromfile(path / "pq_codebooks.bin", dtype=np.float32)
        codebooks_np = codebooks_np.reshape(
            metadata["m"], metadata["k"], metadata["d_sub"]
        )
        quantizer.codebooks.data = torch.from_numpy(codebooks_np).float()
        quantizer.is_trained = True

        return quantizer


class PQDistance(nn.Module):
    """
    Optimized asymmetric distance computation for Product Quantization.

    Computes distances between a query vector and a list of pq encoded
    reference vectors using precomputed distance tables.
    Uses GatherElements optimization for ONNX compatibility.
    """

    def __init__(self):
        super().__init__()

    def forward(
        self,
        query_vector: torch.Tensor,
        reference_codes: torch.Tensor,
        codebooks: torch.Tensor,
    ) -> Tuple[torch.Tensor, torch.Tensor]:
        """
        Compute distance between a query vector and a list of pq encoded reference vectors.

        Args:
            query_vector: Query vector [d]
            reference_codes: PQ codes for reference vectors [N, m]
            codebooks: Codebooks [m, k_centroids, d_sub]

        Returns:
            Tuple of (indices, distances)
        """
        N, m = reference_codes.shape
        _, k_centroids, d_sub = codebooks.shape

        # Reshape query vector into subvectors [m, d_sub]
        query_reshaped = query_vector.view(m, d_sub)

        # Expand query for broadcasting: [m, 1, d_sub]
        query_expanded = query_reshaped.unsqueeze(1)

        # Compute distance tables for all subspaces at once
        # codebooks: [m, k_centroids, d_sub]
        # query_expanded: [m, 1, d_sub]
        diff = query_expanded - codebooks  # [m, k_centroids, d_sub]
        distance_tables = (diff * diff).sum(dim=2)  # [m, k_centroids]

        # Use optimized gathering with proper indexing for ONNX
        # Create batch indices for each subspace
        batch_indices = (
            torch.arange(m, device=reference_codes.device).unsqueeze(0).expand(N, -1)
        )  # [N, m]

        # Create flat indices for GatherElements
        flat_batch_indices = batch_indices.flatten()  # [N*m]
        flat_reference_codes = reference_codes.flatten()  # [N*m]

        # Flatten distance tables for gathering: [m*k_centroids]
        flat_distance_tables = distance_tables.flatten()  # [m*k_centroids]

        # Compute linear indices: batch_idx * k_centroids + code
        linear_indices = flat_batch_indices * k_centroids + flat_reference_codes

        # Gather distances using linear indices
        gathered_distances = flat_distance_tables[linear_indices]  # [N*m]

        # Reshape back to [N, m] and sum across subspaces
        gathered_distances = gathered_distances.view(N, m)  # [N, m]
        total_distances = gathered_distances.sum(dim=1)  # [N]

        # Return all distances and indices sorted (let the caller decide how many to take)
        # This allows merging results from multiple partitions
        top_distances, indices = torch.topk(
            total_distances, N, largest=False, sorted=True
        )

        return indices.to(torch.int32), top_distances


def create_pq_system(
    training_vectors: torch.Tensor,
    m: int = 8,
    k: int = 256,
    max_iterations: int = 100,
    output_dir: Path = Path("."),
) -> Dict:
    """
    Complete PQ system creation: train codebooks and export codebooks and ONNX models.

    Args:
        training_vectors: Training vectors [N, d]
        m: Number of subquantizers
        k: Number of centroids per subquantizer
        max_iterations: Maximum k-means iterations
        output_dir: Directory to save all outputs

    Returns:
        Dictionary with codebooks and metadata
    """
    logger.info("=== Creating Product Quantization System ===")

    d = training_vectors.shape[1]

    # Create, train and save PQ
    pq = PQ(d, m, k)
    pq.train_pq(training_vectors, n_iterations=max_iterations)
    pq.save(output_dir)

    # Export ONNX models
    logger.info(f"Exporting PQ models with d={d}, m={m}, k={k}")

    # Export PQ forward (encode) as ONNX
    torch.onnx.export(
        pq,
        torch.zeros(100, d),
        output_dir / "pq_encode.onnx",
        input_names=["vectors"],
        output_names=["codes"],
        dynamic_axes={"vectors": {0: "n_points"}, "codes": {0: "n_points"}},
        opset_version=13,
        do_constant_folding=True,
        export_params=True,
        dynamo=False,
    )

    # Export PQDistance with optimized distance computation
    distance_model = PQDistance()
    distance_model.eval()

    dummy_query_vector = torch.randn(d)
    dummy_reference_codes = torch.randint(0, k, (1000, m), dtype=torch.uint8)

    torch.onnx.export(
        distance_model,
        (dummy_query_vector, dummy_reference_codes, pq.codebooks.data),
        output_dir / "pq_distance.onnx",
        input_names=["query_vector", "reference_codes", "codebooks"],
        output_names=["indices", "distances"],
        dynamic_axes={
            "reference_codes": {0: "n_references"},
            "indices": {0: "k_out"},
            "distances": {0: "k_out"},
        },
        opset_version=13,
        do_constant_folding=True,
        export_params=True,
        dynamo=False,
    )

    logger.info(f"Exported PQ models to {output_dir}")
    logger.info("- pq_encode.onnx: Encode vectors to PQ codes")
    logger.info(
        "- pq_distance.onnx: Optimized asymmetric distance computation (returns all distances sorted)"
    )

    result = {
        "codebooks": pq.codebooks.data,
        "metadata": {
            "d": pq.d,
            "m": pq.m,
            "k": pq.k,
            "d_sub": pq.d_sub,
            "compression_ratio": float((pq.d * 32) / (pq.m * 8)),
            "training_samples": training_vectors.shape[0],
            "max_iterations": max_iterations,
            "version": "1.0",
        },
    }

    logger.info("=== PQ System Creation Complete ===")
    return result
